flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java

```diff
@@ -0,0 +1,488 @@
+	public final OperatorSnapshotFutures snapshotState(long checkpointId, long timestamp, CheckpointOptions checkpointOptions,
+			CheckpointStreamFactory factory) throws Exception {
+		return stateHandler.snapshotState(
+			this::snapshotState,
+			getOperatorName(),
+			checkpointId,
+			timestamp,
+			checkpointOptions,
+			factory);
+	}
+
+	/**
+	 * Stream operators with state, which want to participate in a snapshot need to override this hook method.
+	 *
+	 * @param context context that provides information and means required for taking a snapshot
+	 */
+	public void snapshotState(StateSnapshotContext context) throws Exception {
+	}
+
+	/**
+	 * Stream operators with state which can be restored need to override this hook method.
+	 *
+	 * @param context context that allows to register different states.
+	 */
+	public void initializeState(StateInitializationContext context) throws Exception {
+
+	}
+
+	@Override
+	public void notifyCheckpointComplete(long checkpointId) throws Exception {
+		stateHandler.notifyCheckpointComplete(checkpointId);
+	}
+
+	// ------------------------------------------------------------------------
+	//  Properties and Services
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Gets the execution config defined on the execution environment of the job to which this
+	 * operator belongs.
+	 *
+	 * @return The job's execution config.
+	 */
+	public ExecutionConfig getExecutionConfig() {
+		return executionConfig;
+	}
+
+	public StreamConfig getOperatorConfig() {
+		return config;
+	}
+
+	public ClassLoader getUserCodeClassloader() {
+		return userCodeClassLoader;
+	}
+
+	/**
+	 * Return the operator name. If the runtime context has been set, then the task name with
+	 * subtask index is returned. Otherwise, the simple class name is returned.
+	 *
+	 * @return If runtime context is set, then return task name with subtask index. Otherwise return
+	 * 			simple class name.
+	 */
+	protected String getOperatorName() {
+		if (runtimeContext != null) {
+			return runtimeContext.getTaskNameWithSubtasks();
+		} else {
+			return getClass().getSimpleName();
+		}
+	}
+
+	/**
+	 * Returns a context that allows the operator to query information about the execution and also
+	 * to interact with systems such as broadcast variables and managed state. This also allows
+	 * to register timers.
+	 */
+	public StreamingRuntimeContext getRuntimeContext() {
+		return runtimeContext;
+	}
+
+	@SuppressWarnings("unchecked")
+	public <K> KeyedStateBackend<K> getKeyedStateBackend() {
+		return (KeyedStateBackend<K>) stateHandler.getKeyedStateBackend();
+	}
+
+	public OperatorStateBackend getOperatorStateBackend() {
+		return stateHandler.getOperatorStateBackend();
+	}
+
+	/**
+	 * Returns the {@link ProcessingTimeService} responsible for getting the current
+	 * processing time and registering timers.
+	 */
+	public ProcessingTimeService getProcessingTimeService() {
+		return processingTimeService;
+	}
```

Comment:
Reducing visibility would fail migration of only 2 tests.
We don't know though when this will happen. And then we could either fix the tests, disable them temporarily, or change visibility back to `public`.
So I think it's better to have it protected until then.

Similar argument applies to `numEventTimeTimers` and `numProcessingTimeTimers`.

Reply:
As I mentioned above, I would prefer to keep it `public`, but we can mark those methods `@VisibleForTesting`?

Reply:
I'd rather go with `protected` but I don't think this issue is very important.